design-principles.md

```diff
@@ -0,0 +1,26 @@
+# Tekton Design Principles
+
+## Reusability
+- Existing features should be reused when possible instead of adding new ones. Before proposing a feature, try solving the problem with existing features first.
+- Prefer providing a solution in the Catalog when possible to adding a feature.
+- In design docs, discuss how a new feature affects the reusability of Tasks and Pipelines.
+
+## Simplicity 
+- Tekton should contain only the bare minimum and simplest features needed to meet the largest number of CI/CD use cases.
+- Prefer a [simple](https://www.infoq.com/presentations/Simple-Made-Easy/) solution that solves most use cases to a complex solution that solves all use cases (can be revisited later).
+- New features should be consistent with existing components, in structure and behavior, to make learnability, trialability and adoption easy.
+- Any new feature should have been previously discussed and agreed upon in a [Tekton Enhancement Proposal](https://github.com/tektoncd/community/tree/master/teps). 
+- In design docs, demonstrate that the proposed feature is absolutely necessary. What’s the current experience without the feature and how challenging is it?
+
+## Flexibility
+- Tekton has a ton of flexibility, which means a lot of things can be implemented by some kind of plugin, such as using `CustomTasks` or in the `Step` level. When considering adding something to Tekton itself, we should consider and exhaust all opportunities to implement it using one of the existing plugin mechanisms.
+- Specific choices should be implemented at the step or DSL levels by the users, not at the Task or Pipeline levels to keep Tekton flexible. Such as Tekton should not be aware of git concepts, such as branches. 
```

Comment:
I agree with this, but the step vs task/pipeline distinction is not 100% clear to me.
Does this mean for example that we want a `CloneRepo` task that includes branch logic in the step, `if git then elif snv then ...`? 
I think having tasks specifically designed and tested for each SCM, selected with conditionals (when expressions?_) etc would be a valid alternative.

Reply:
that makes sense to me, maybe we can reword this as something like:

- To keep Tekton flexible, specific choices should be implemented within Tasks or at the DSL level by users, not at the level of features provided by Tekton in the Task or Pipeline APIs. Such as Tekton should not be aware of git concepts, such as branches.

Reply:
@jerop thanks for the update. 
Sorry to insist on this, but I feel it would be better to be specific about the Tekton API, as @bobcatfish suggested. 
Tekton should avoid being opinionated in the Task and Pipeline *API* , and Tasks (e.g. from the catalog) should be a valid place where to be specific and opinionated.

Reply:
thanks for making that point clearer, updated it: https://github.com/tektoncd/community/pull/171/files#diff-0899720b7e7bf227aaa00d838f2c9e76R17